<table align="left" width=100%>
    <tr>
        <td>
            <div align="center">
                <font color="#21618C" size=8px>
                  <b> Faculty Notebook </b>
                </font>
            </div>
        </td>
    </tr>
</table>

## Table of Content

1. **[Import Libraries](#lib)**
2. **[Parameter Estimation](#est)**
    - 2.1 - **[Point Estimation](#pt)**
        - 2.1.1 - **[Sampling Error](#err)**
    - 2.2 - **[Interval Estimation for Mean](#int)**
        - 2.2.1 - **[Large Sample Size](#large)**   
        - 2.2.2 - **[Small Sample Size](#small)**
    - 2.3 - **[Interval Estimation for Proportion](#prop)**
3. **[Test of Hypothesis](#defn)**
    - 3.1 - **[Types of Hypothesis](#types)**
4. **[Types of Test](#test_type)**
    - 4.1 - **[Two Tailed Test](#2tailed)**
    - 4.2 - **[One Tailed Test](#1tailed)**
5. **[Hypothesis Tests with Z Statistic](#eg)**
6. **[Small Sample Test](#t)**
    - 6.1 - **[One Sample t Test](#1t)**
7. **[Error](#error)**

<a id="lib"></a>
# 1. Import Libraries

**Let us import the required libraries.**

In [87]:
# import 'pandas'
import pandas as pd

# import 'numpy'
import numpy as np

# import subpackage of matplotlib
import matplotlib.pyplot as plt

# import 'seaborn'
import seaborn as sns

# to suppress warnings
from warnings import filterwarnings
filterwarnings('ignore')

# import 'random' to generate random sample
import random

# import statistics to perform statistical computation
import statistics

# import 'stats' package from scipy library
from scipy import stats

# import a library to perform Z-test
from statsmodels.stats import weightstats as stests

# to test the normality
from scipy.stats import shapiro

# import the function to calculate the power of test
from statsmodels.stats import power

In [ ]:
# set the plot size using 'rcParams'
# once the plot size is set using 'rcParams', it sets the size of all the forthcoming plots in the file
# pass width and height in inches to 'figure.figsize'
plt.rcParams['figure.figsize'] = [15,8]

<a id="est"></a>
# 2. Parameter Estimation

The value associated with the characteristic of the population is known as a `parameter` and the characteristic of the sample is described by a `statistic`.
Usually, the population parameters are not known in most of the real-life problems. Thus, we consider a subset of the population (sample) to estimate the population parameter using a sample statistic.

`Point estimation` and `Interval estimation` are two of the methods to estimate the population parameter.

<a id="pt"></a>
## 2.1 Point Estimation

This method considers a single value (sample statistic) as estimate of the population parameter.

Let $X_{1}, X_{2}, X_{3},..., X_{n}$ be the random sample drawn from a population with mean $\mu$ and standard deviation $\sigma$. <br>
The point estimation method estimates the population mean, $\mu = \overline{X}$, where $\overline{X}$ is the sample mean and population standard deviation, $\sigma = s$, where $s$ is the standard deviation of the sample (`Standard Error`).

### Example:

#### 1. Consider the data of grade points for 35 students in a data science course. Select grades of 20 students randomly from the data and find the point estimate for the population mean.

     Grades: [59.1, 65.0, 75.8, 79.2, 95.0, 99.8, 89.1, 65.2, 41.9, 55.2, 94.8, 84.1, 83.2, 74.0, 75.5, 76.2, 79.1, 80.1,
              92.1, 74.2, 59.2, 64.0, 75, 78.2, 95.6, 97.8, 89.5, 64.2, 41.8, 57.2, 85, 91.4, 81.8, 74.6, 90]

In [9]:
Grades= [59.1, 65.0, 75.8, 79.2, 95.0, 99.8, 89.1, 65.2, 41.9, 55.2, 94.8, 84.1, 83.2, 74.0, 75.5, 76.2, 79.1, 80.1,
          92.1, 74.2, 59.2, 64.0, 75, 78.2, 95.6, 97.8, 89.5, 64.2, 41.8, 57.2, 85, 91.4, 81.8, 74.6, 90]

In [29]:
import random
import numpy as np

x=random.sample(population=Grades,k=20)
round(np.mean(x),2)

74.74

In [2]:
# given population
grades = [59.1, 65.0, 75.8, 79.2, 95.0, 99.8, 89.1, 65.2, 41.9, 55.2, 94.8, 84.1, 83.2, 74.0, 75.5, 76.2, 79.1, 80.1,
          92.1, 74.2, 59.2, 64.0, 75, 78.2, 95.6, 97.8, 89.5, 64.2, 41.8, 57.2, 85, 91.4, 81.8, 74.6, 90]

# draw a sample for grades of 18 students without replacement
# set a random seed to get the same sample each time you run the code
random.seed(1)

# pass the data to the parameter, 'population'
# pass the required sample size to the parameter, 'k'
samp_grades = random.sample(population = grades, k = 20)

# the point estimate for the population mean is equal to 'sample mean'
# calculate the sample mean
samp_mean = np.mean(samp_grades)

print('The point estimate for the population mean is', samp_mean)

The point estimate for the population mean is 75.32500000000002


#### 2. A financial firm has created 50 portfolios. From them, a sample of 13 portfolios was selected, out of which 8 were found to be underperforming. Can you estimate the number of underperforming portfolios in the population?

In [35]:
(1500*10)/100

150.0

In [39]:
n=13
x=8
pop=50
y=8/13
y
50*y

30.76923076923077

In [3]:
# total count of portfolios
N = 50

# number of portfolios in a sample
n = 13

# number of underperforming portfolios in a sample
x = 8

# sample proportion
p_samp = x/n

# estimate the number of underperforming portfolios
num_port = p_samp*N

# round the number to get an integer value
print('The number of underperforming portfolios:', round(num_port))

The number of underperforming portfolios: 31


<a id="err"></a>
### 2.1.1 Sampling Error

Sampling error is considered as the absolute difference between the sample statistic used to estimate the parameter and the corresponding population parameter. Since the entire population is not considered as the sample, the values of mean, median, quantiles, and so on calculated on sample differ from the actual population values.

One can reduce the sampling error either by increasing the sample size or determining the optimal sample size using various methods.

### Example:

#### 1. Consider the data for the number of ice-creams sold per day. An ice-cream vendor collected this data for 90 days and then a sample is drawn (without replacement) containing ice-creams sold for 25 days.

data = [21, 93, 62, 76, 73, 20, 56, 95, 41, 36, 38, 13, 80, 88, 34, 18, 40, 11,
        25, 29, 61, 23, 82, 10, 92, 69, 60, 87, 14, 91, 94, 49, 57, 83, 96, 55,
        79, 52, 59, 39, 58, 17, 19, 98, 15, 54, 48, 46, 72, 45, 65, 28, 37, 30,
        68, 75, 16, 33, 31, 99, 22, 51, 27, 67, 85, 47, 44, 77, 64, 97, 84, 42,
        90, 70, 74, 89, 32, 26, 24, 12, 81, 53, 50, 35, 71, 63, 43, 86, 78, 66]
        
sample = [10, 22, 47, 66, 11, 57, 77, 98, 31, 63, 74, 84, 50, 96, 88, 92, 70, 54, 65, 44, 16, 72, 20, 90, 43]

Comupte the sampling error for mean.

In [79]:
data = [21, 93, 62, 76, 73, 20, 56, 95, 41, 36, 38, 13, 80, 88, 34, 18, 40, 11, 25, 29, 61, 23, 82, 10, 92, 69, 60, 87, 14, 91, 94, 49, 57, 83, 96, 55, 79, 52, 59, 39, 58, 17, 19, 98, 15, 54, 48, 46, 72, 45, 65, 28, 37, 30, 68, 75, 16, 33, 31, 99, 22, 51, 27, 67, 85, 47, 44, 77, 64, 97, 84, 42, 90, 70, 74, 89, 32, 26, 24, 12, 81, 53, 50, 35, 71, 63, 43, 86, 78, 66]
n=np.random.choice(data,25)
pop_mean=np.mean(data)
sam_mean=np.mean(n)

In [81]:
abs(sam_mean-pop_mean)

5.460000000000001

In [ ]:
# given population
data = [21, 93, 62, 76, 73, 20, 56, 95, 41, 36, 38, 13, 80, 88, 34, 18, 40, 11, 25, 29, 61, 23, 82, 10, 92, 69, 60, 87, 14, 91,
        94, 49, 57, 83, 96, 55, 79, 52, 59, 39, 58, 17, 19, 98, 15, 54, 48, 46, 72, 45, 65, 28, 37, 30, 68, 75, 16, 33, 31, 99,
        22, 51, 27, 67, 85, 47, 44, 77, 64, 97, 84, 42, 90, 70, 74, 89, 32, 26, 24, 12, 81, 53, 50, 35, 71, 63, 43, 86, 78, 66]

# given sample
sample = [10, 22, 47, 66, 11, 57, 77, 98, 31, 63, 74, 84, 50, 96, 88, 92, 70, 54, 65, 44, 16, 72, 20, 90, 43]

# calculate the point estimate for the population mean
samp_mean = np.mean(sample)
print(samp_mean)

# population mean
pop_mean = np.mean(data)
print(pop_mean)

# calculate the sampling error for mean
print("Sampling error for mean:", np.abs(samp_mean - pop_mean).round(3))

57.6
54.5
Sampling error for mean: 3.1


The average temperature on a summer morning in Alaska is 55oF. The research student of
climate studies records the temperature for 15 summer mornings in oF.
Data = [49.8, 52.3, 51, 56.9, 54.8, 63, 58.2, 54.1, 50.4, 49.2, 47, 51.3, 43.5, 56, 55]
Find the sampling error for the average temperature.

In [ ]:
import numpy as np
Data = [49.8, 52.3, 51, 56.9, 54.8, 63, 58.2, 54.1, 50.4, 49.2, 47, 51.3, 43.5, 56, 55]
pop_mean=55
samp_mean=np.mean(Data)
print(samp_mean)
print('Sampling error for mean:',np.abs(samp_mean-pop_mean))

52.833333333333336
Sampling error for mean: 2.1666666666666643


<a id="int"></a>
## 2.2 Interval Estimation for Mean

This method considers the range of values in which the population parameter is likely to lie. The confidence interval is an interval that describes the range of values in which the parameter lies with a specific probability. It is given by the formula,<br> <p style='text-indent:20em'> `conf_interval = sample statistic ± margin of error`</p>

The uncertainty of an estimate is described by the `confidence level` which is used to calculate the margin of error.

<a id="large"></a>
### 2.2.1 Large Sample Size

Consider a population with mean $\mu$ and standard deviation $\sigma$. Let us take a sample of `n` observations from the population such that, $n \geq 30$. The central limit theorem states that the sampling distribution of mean follows a normal distribution with mean $\mu$ and standard deviation $\frac{\sigma}{\sqrt(n)}$.

The confidence interval for the population mean with $100(1-\alpha)$% confidence level is given as: $\overline{X} \pm Z_{\frac{\alpha}{2}}\frac{\sigma}{\sqrt{n}}$

Where, <br>
$\overline{X}$: Sample mean<br>
$\alpha$: Level of significance<br>
$\sigma$: Population standard deviation<br>
$n$: Sample size

The quantity $\frac{\sigma}{\sqrt{n}}$ is the standard error of the mean. And the margin of error is given by $Z_{\frac{\alpha}{2}}\frac{\sigma}{\sqrt{n}}$.

If we know the expected margin of error (ME), then we can calculate the required sample size (n) using the formula: $n = (Z_{\frac{\alpha}{2}})^{2}\frac{\sigma^{2}}{ME^{2}}$.

The above equation is valid for any population provided the sample size is sufficiently large (usually $n \geq 30$). Relace $\sigma$ by the standard deviation of the sample ($s$) if the population standard deviation is not known.

The value of $Z_{\frac{\alpha}{2}}$ for different $\alpha$ values can be obtained using the `stats.norm.isf()` from the scipy library.

In [ ]:
# let us find the Z-values for different alpha values

# create an empty dataframe to store the alpha and corresponding Z-value
df_Z = pd.DataFrame()

# create a dictionary of different alpha values
alpha =  [0.1, 0.05, 0.02, 0.01]

# use for loop to calculate the value for each alpha
for i in range(len(alpha)):
    alpha_by_2 = alpha[i] / 2

    # use 'stats.norm.isf()' to find the Z-value corresponding to the upper tail probability 'q'
    # pass the value of 'alpha_by_2' to the parameter 'q'
    # use 'round()' to round-off the value to 4 digits
    Z = np.abs(round(stats.norm.isf(q = alpha_by_2), 4))

    # create a dataframe using dictionary to store the alpha and corresponding Z-value
    # set the loop iterator 'i' as the index of the dataframe
    row =  pd.DataFrame({"alpha": alpha[i], "Z_alpha_by_2" : Z}, index = [i])

    # append the row to the dataframe 'df_Z'
    df_Z = df_Z.append(row)

# print the final dataframe
df_Z

To calculate the confidence interval with 95% confidence, use the Z-value corresponding to `alpha = 0.05`.

### Example:

#### 1. A random sample of weight (in kg.) for 35 diabetic patients is drawn from the population with a standard deviation of 8 kg. Find the 90% confidence interval for the population mean.

    Weight: [59.1, 65.0, 75.8, 79.2, 95.0, 99.8, 89.1, 65.3, 41.9, 55.2, 94.8, 84.1, 83.2, 74.0, 75.5, 76.2, 79.1, 80.1,
             92.1, 74.2, 59.2, 64.0, 75, 78.2, 95.6, 97.8, 89.5, 64.2, 41.8, 57.2, 85, 91.4, 81.8, 74.6, 90]

In [85]:
weight = [59.1, 65.0, 75.8, 79.2, 95.0, 99.8, 89.1, 65.3, 41.9, 55.2, 94.8, 84.1, 83.2, 74.0, 75.5, 76.2, 79.1, 80.1,
          92.1, 74.2, 59.2, 64.0, 75, 78.2, 95.6, 97.8, 89.5, 64.2, 41.8, 57.2, 85, 91.4, 81.8, 74.6, 90]

x=len(weight)
avg=np.mean(weight)
std=8
z_alpha_2=1.64
lower_tail=avg-(z_alpha_2*(std/np.sqrt(x)))
upper_tail=avg+(z_alpha_2*(std/np.sqrt(x)))
lower_tail  ,upper_tail            

(74.46802952130666, 78.9033990501219)

In [ ]:
# given sample
weight = [59.1, 65.0, 75.8, 79.2, 95.0, 99.8, 89.1, 65.3, 41.9, 55.2, 94.8, 84.1, 83.2, 74.0, 75.5, 76.2, 79.1, 80.1,
          92.1, 74.2, 59.2, 64.0, 75, 78.2, 95.6, 97.8, 89.5, 64.2, 41.8, 57.2, 85, 91.4, 81.8, 74.6, 90]

# number of observations in sample
n = len(weight)

# given population standard deviation
std_pop = 8

# calculate the 90% confidence interval
# pass the sample mean to the parameter, 'loc'
# pass the scaling factor (std_pop / n^(1/2)) to the parameter, 'scale'
interval = stats.norm.interval(0.90, loc = np.mean(weight), scale = std_pop / np.sqrt(n))

# print the 90% confidence interval
# use 'round()' to round-off the value to 2 digits
print('The 90% confidence interval of population mean is', np.round(interval,2))

The 90% confidence interval of population mean is [74.46 78.91]


#### Let us calculate the confidence interval without using the built-in function.

In [ ]:
# given sample
weight = [59.1, 65.0, 75.8, 79.2, 95.0, 99.8, 89.1, 65.2, 41.9, 55.2, 94.8, 84.1, 83.2, 74.0, 75.5, 76.2, 79.1, 80.1,
          92.1, 74.2, 59.2, 64.0, 75, 78.2, 95.6, 97.8, 89.5, 64.2, 41.8, 57.2, 85, 91.4, 81.8, 74.6, 90]

# calculate sample mean
avg  = np.mean(weight)

# given population standard deviation
std_pop = 8

# number of observations in sample
n = len(weight)

# Z-value for 90% confidence level
z_alpha_by_2 = 1.6449

# calculate the lower and upper bound of interval using the formula
lower_int = avg - (z_alpha_by_2 * (std_pop / np.sqrt(n)))
upper_int = avg + (z_alpha_by_2 * (std_pop / np.sqrt(n)))

# print the 90% confidence interval
# use 'round()' to round-off the value to 2 digits
print('90% confidence interval for population mean is', (round(lower_int,2), round(upper_int,2)))

90% confidence interval for population mean is (74.46, 78.91)


#### 2. There are 150 apples on a tree. You randomly choose 40 apples and found that the average weight of apples is 182 grams with a standard deviation of 30 grams. Find the 95% confidence interval for the population mean.

In [ ]:
# number of observations in the sample
n = 40

# sample mean
sample_avg = 182

# sample standard deviation
sample_std = 30

# calculate the 95% confidence interval
# pass the sample mean to the parameter, 'loc'
# pass the scaling factor (sample_std / n^(1/2)) to the parameter, 'scale'
# as the population standard deviation is unknown, use the sample standard deviation
interval = stats.norm.interval(0.95, loc = sample_avg, scale = sample_std / np.sqrt(n))
print('95% confidence interval for population mean is', np.round(interval,2))

95% confidence interval for population mean is [172.7 191.3]


#### 3. A movie production house needs to estimate the average monthly wage of the technical crew members. The previous data shows that the standard deviation of the wages is 190 dollars. The production team thinks that the estimation of the average wage should not exceed 54 dollars. The team has decided to take a small subset of wages for the estimation. Find a suitable number of wages to be considered to get the estimate with 90% confidence.

In [ ]:
# calculate sample size using the margin of error

# given population standard deviation
sigma = 190

# required margin of error
ME = 54

# calculate z_alpha_by_2 with alpha = 0.1
# use 'stats.norm.isf()' to find the Z-value corresponding to the upper tail probability 'q'
# pass the value of 'alpha/2' to the parameter 'q'
# use 'round()' to round-off the value to 4 digits
z_alpha_by_2 = np.abs(round(stats.norm.isf(q = 0.1/2), 4))

# calculate sample size (n)
n = ((z_alpha_by_2)**2)*(sigma**2)/(ME**2)

# round-off the value of 'n' to get the integer value
print('Required Sample Size:', round(n))

Required Sample Size: 33


#### 4. 100 bags of coal were tested and had an average of 35% of ash with a standard deviation of 15%. Calculate the margin of error for a 95% confidence level.

In [ ]:
# number of bags
n = 100

# standard deviaion
std = 0.15

# given confidence level
conf_level = 0.95

# calculate z_alpha_by_2 with alpha = (1-conf_level) = 0.05
# use 'stats.norm.isf()' to find the Z-value corresponding to the upper tail probability 'q'
# pass the value of 'alpha/2' to the parameter 'q'
# use 'round()' to round-off the value to 4 digits
z_alpha_by_2 = np.abs(round(stats.norm.isf(q = 0.05/2), 4))

# calculate margin of error
error = (z_alpha_by_2*std)/ np.sqrt(n)

print('Margin of error:', error)

Margin of error: 0.0294


#### 5. From a sample of 250 observations, it is found that the average income of a 27 year old Londoner is £45,000 with a sample standard deviation of £4000. Obtain the 95% confidence interval to estimate the average income.

In [ ]:
n=250
avg=45000
std=4000
ci=0.95
z_alpha_by_2=round(stats.norm.isf(q=0.05/2),2)
print(z_alpha_by_2)
upp_int=avg+(z_alpha_by_2*(std/np.sqrt(n)))
low_int=avg-(z_alpha_by_2*(std/np.sqrt(n)))
print(upp_int,low_int)

1.96
45495.845137114404 44504.154862885596


In [ ]:
# number of observations in the sample
n = 250

# sample mean
sample_avg = 45000

# sample standard deviation
sample_std = 4000

# calculate the 95% confidence interval
# pass the sample mean to the parameter, 'loc'
# pass the scaling factor (sample_std / n^(1/2)) to the parameter, 'scale'
# as the population standard deviation is unknown, use the sample standard deviation
interval = stats.norm.interval(0.95, loc = sample_avg, scale = sample_std/np.sqrt(n))

print('95% confidence interval for average income:', np.round(interval,2))

95% confidence interval for average income: [44504.16 45495.84]


<a id="small"></a>
### 2.2.2 Small Sample Size

Let us take a sample of `n` observations from the population such that, $n < 30$. Here the standard deviation of the population is unknown. The confidence interval for the population mean with $100(1-\alpha)$% confidence level is given as: $\overline{X} \pm t_{\frac{\alpha}{2}, n-1}\frac{s}{\sqrt{n}}$

Where, <br>
$\overline{X}$: Sample mean<br>
$\alpha$: Level of significance<br>
$s$: Sample standard deviation<br>
$n-1$: degrees of freedom

The ratio $\frac{s}{\sqrt{n}}$ is the estimate of the standard error of the mean. And $t_{\frac{\alpha}{2}, n-1}\frac{s}{\sqrt{n}}$ is the margin of error for the estimate.

The value of $t_{\frac{\alpha}{2}, n-1}$ for different $\alpha$ values can be obtained using the `stats.t.isf()` from the scipy library.  

In [ ]:
# let us find the t-values for different alpha values with 10 degrees of freedom

# create an empty dataframe to store the alpha and corresponding t-value
df_t = pd.DataFrame()

# create a dictionary of different alpha values
alpha =  [0.1, 0.05, 0.02, 0.01]

# use for loop to calculate the t-value for each alpha value
for i in range(len(alpha)):
    alpha_by_2 = alpha[i] / 2

    # use 'stats.t.isf()' to find the t-value corresponding to the upper tail probability 'q'
    # pass the value of 'alpha_by_2' to the parameter 'q'
    # pass the 10 degrees of freedom to the parametr 'df'
    # use 'round()' to round-off the value to 2 digits
    t = np.abs(round(stats.t.isf(q = alpha_by_2, df = 10), 2))

    # create a dataframe using dictionary to store the alpha and corresponding t-value
    # set the loop iterator 'i' as the index of the dataframe
    row =  pd.DataFrame({"alpha": alpha[i], "t_alpha_by_2": t}, index = [i])

    # append the row to the dataframe 'df_t'
    df_t = df_t.append(row)

# print the final dataframe
df_t

### Example:

#### 1. There are 150 apples on a tree. You randomly choose 17 apples and found that the average weight of apples is 78 grams with a standard deviation of 23 grams. Find the 90% confidence interval for the population mean.

In [ ]:
# number of observations in the sample
n = 17

# sample mean
sample_avg = 78

# sample standard deviation
sample_std = 23

# calculate the 90% confidence interval for the population mean
# pass the 'n-1' degrees of freedom to the parameter, 'df'
# pass the sample mean to the parameter, 'loc'
# pass the scaling factor (sample_std / n^(1/2)) to the parameter, 'scale'
# as the population standard deviation is unknown, use the sample standard deviation
interval = stats.t.interval(0.90, df = n-1, loc = sample_avg, scale = sample_std/np.sqrt(n))

print('90% confidence interval for population mean is', np.round(interval,2))

90% confidence interval for population mean is [68.26 87.74]


<a id="prop"></a>
## 2.3 Interval Estimation for Proportion

Consider a population in which each observation is either a success or a failure. The population proportion is denoted by `P` which the ratio of the number of successes to the size of the population.

The confidence interval for the population proportion with $100(1-\alpha)$% confidence level is given as: $p \pm Z_{\frac{\alpha}{2}}\sqrt{\frac{p(1 - p)}{n}}$

Where, <br>
$p$: Sample proportion<br>
$\alpha$: Level of significance<br>
$n$: Sample size

The quantity $Z_{\frac{\alpha}{2}}\sqrt{\frac{p(1 - p)}{n}}$ is the margin of error.

### Example:

#### 1. A financial firm has created 50 portfolios. From them, a sample of 13 portfolios was selected, out of which 8 were found to be underperforming. Construct a 99% confidence interval to estimate the population proportion.

In [ ]:
# total count of portfolios
N = 50

# number of portfolios in a sample
n = 13

# number of underperforming portfolios in a sample
x = 8

# sample proportion
p_samp = x/n

# calculate the 99% confidence interval
# pass the sample proportion to the parameter, 'loc'
# pass the scaling factor ((p_samp*(1-p_samp))/n))^0.5) to the parameter, 'scale'
interval = stats.norm.interval(0.99, loc = p_samp, scale = np.sqrt((p_samp*(1-p_samp))/n))

print('99% confidence interval for population proportion is', np.round(interval,2))

<a id="defn"></a>
# 3. Test of Hypothesis

Hypothesis testing is the process of evaluating the validity of claims made about the population using the sample data obtained from the population. A statistical test is a rule used to decide whether to reject or retain the claim.

**Examples of hypothesis:**

        1. One can get 'A' grade if the attendance in the class is more than 75%.
        2. A probiotic drink can improve the immunity of a person.

<a id="types"></a>
## 3.1 Types of Hypothesis

`Null Hypothesis`: The null hypothesis is the statistical hypothesis suggesting 'no difference' between the population parameter and a specific value
. It is denoted as H<sub>0</sub>.

`Alternative Hypothesis`: It is the hypothesis that is tested against the null hypothesis and states the existence of a difference between the parameter and a specific value. It is denoted by H<sub>a</sub> or H<sub>1</sub>.

The claim is usually the alternative hypothesis H<sub>1</sub> also known as research hypothesis. To test the claim we need to gather evidence (data ) and find the likelihood of the data under the assumption H<sub>0</sub> being true.

#### Example:

A company that produces tennis balls claimed that the diameter of a tennis ball is at least 2.625 inches on average. On the other hand, a professional tennis coach claimed that the diameter of a ball is less than what the company has claimed. To test the claim of the coach, a statistical test can be performed considering the hypothesis:

                    Null Hypothesis: Average diameter ≥ 2.625
                    Alternative Hypothesis: Average diameter < 2.625

<a id="test_type"></a>
# 4. Types of Test

The hypothesis test is used to validate the claims about the population. The types of tests are based on the nature of the alternative hypothesis.

<a id="2tailed"></a>
## 4.1 Two Tailed Test

Two tailed test considers the value of the population parameter is less than or greater than (i.e. not equal) a specific value. <br>
If we test the population mean ($\mu$) with a specific value ($\mu_{0}$) the null hypothesis is: $H_{0}: \mu = \mu_{0}$.

The alternative hypothesis for the two tailed test is given as: $H_{1}: \mu \neq \mu_{0}$

#### Example:

A company that produces tennis balls claimed that the diameter of a tennis ball is 2.625 inches on average. To test the company's claim, a statistical test can be performed considering the hypothesis:







                    Null Hypothesis: Average diameter = 2.625
                    Alternative Hypothesis: Average diameter ≠ 2.625

<a id="1tailed"></a>
## 4.2 One Tailed Test

One tailed test considers the value of the population parameter is less than or greater than (but not both) a specific value. <br>
If we test the population mean ($\mu$) with a specific value ($\mu_{0}$) the null hypothesis is: $H_{0}: \mu \leq \mu_{0}$ and the alternative hypothesis is $H_{1}: \mu > \mu_{0}$, the one tailed test is also known as a `right-tailed test`.

If we test the population mean ($\mu$) with a specific value ($\mu_{0}$) the null hypothesis is: $H_{0}: \mu \geq \mu_{0}$ and the alternative hypothesis is $H_{1}: \mu < \mu_{0}$, the one tailed test is also known as a `left-tailed test`.


### Example:

**1.** The company's annual quality report of machines states that a lathe machine works efficiently at most for 8 months on average after the servicing. The production manager claims that after the special tuxan servicing, the machine works efficiently for more than 8 months. To test the claim of production manager consider the hypothesis:














                    Null Hypothesis: Machine efficiency ≤ 8 months
                    Alternative Hypothesis: Machine efficiency > 8 months

This is the example of a **right-tailed test**.

**2.** A railway authority claims that all the trains on the Chicago-Seattle route run with a speed of at least 54 mph on average. A customer forum declares that there are various records from passengers claiming that the speed of the train is less than what railway has claimed. In this scenario, a statistical test can be performed to test the claim of customer forum considering the hypothesis:










                    Null Hypothesis: Speed ≥ 56 mph
                    Alternative Hypothesis: Speed < 56 mph

This is the example of a **left-tailed test**.

<a id="eg"></a>
# 5. Hypothesis Tests with Z Statistic

Let us perform one sample Z test for the population mean. We compare the population mean with a specific value. The sample is assumed to be taken from a population following a normal distribution.

To check the normality of the data, a test for normality is used. The `Shapiro-Wilk Test` is one of the methods used to check the normality. The hypothesis of the test is given as:
<p style='text-indent:25em'> <strong> H<sub>0</sub>:  The data is normally distributed </strong> </p>
<p style='text-indent:25em'> <strong> H<sub>1</sub>:  The data is not normally distributed </strong> </p>

The `shapiro()` from scipy library performs a Shapiro-Wilk normality test.

The null and alternative hypothesis of Z-test is given as:
<p style='text-indent:25em'> <strong> $H_{0}: \mu = \mu_{0}$ or $\mu \geq \mu_{0}$ or $\mu \leq \mu_{0}$</strong></p>
<p style='text-indent:25em'> <strong> $H_{1}: \mu \neq \mu_{0}$ or $\mu < \mu_{0}$ or $\mu > \mu_{0}$</strong></p>

Consider a normal population with standard deviation $\sigma$. Let us take a sample of size n, such that (n > 30).
The test statistic for one sample Z-test is given as:
<p style='text-indent:25em'> <strong> $Z = \frac{\overline{X} -  \mu}{\frac{\sigma}{\sqrt(n)}}$</strong></p>

Where, <br>
$\overline{X}$: Sample mean<br>
$\mu$: Specified mean<br>
$\sigma$: Population standard deviation<br>
$n$: Sample size

Under $H_{0}$ the test statistic follows a standard normal distribution.

If $\sigma$ is unknown, use the sample standard deviation (s) instead of $\sigma$ to calculate the test statistic.

In [ ]:
z_alpha_by_2=0.1/2=0.05

### Example:

#### 1. A car manufacturing company claims that the mileage of their new car is 25 kmph with a standard deviation of 2.5 kmph. A random sample of 45 cars was drawn and recorded their mileage as per the standard procedure. From the sample, the mean mileage was seen to be 24 kmph. Is this evidence to claim that the mean mileage is different from 25kmph? (assume the normality of the data) Use α = 0.01.

In [ ]:
two tailed test--->
h0:mileage=25
h1:mileage!=25

In [ ]:
pop_mean=25
pop_std=2.5
n=45
samp_mean=24
confidence_value=0.01

In [ ]:
n=45
pop_std=2.5
pop_mean=25
samp_mean=24
confidence_interval=0.01

In [91]:
z_alpha_2=round(stats.norm.isf(q=0.01/2),2)
z_alpha_2

2.58

In [99]:
def z_score(pop_mean,pop_std,samp_mean,n):
    z_test=(samp_mean-pop_mean)/(pop_std/np.sqrt(n))
    return z_test
n=45
pop_std=2.5
pop_mean=25
samp_mean=24
confidence_interval=0.01

x=z_score(pop_mean,pop_std,samp_mean,n)
print(x)

-2.6832815729997477


In [ ]:
SAMP_MEAN=24
POP_MEAN=25
POP_STD=2.5
n=45
confidence_level=99%

The null and alternative hypothesis is:

H<sub>0</sub>: $\mu = 25 $<br>
H<sub>1</sub>: $\mu ≠ 25 $

Here ⍺ = 0.01, for a two-tailed test calculate the critical z-value.

In [8]:
abs(-0.25)

0.25

In [20]:
# calculate the z-value for 99% of confidence level
# use 'stats.norm.isf()' to find the z-value corresponding to the upper tail probability 'q'
# pass the value of 'alpha/2' to the parameter 'q', here alpha = 0.01
# use 'round()' to round-off the value to 2 digits
z_val = np.abs(round(stats.norm.isf(q = 0.01/2), 2))

print('Critical value for two-tailed Z-test:', z_val)

Critical value for two-tailed Z-test: 2.58


i.e. if z is less than -2.58 or greater than 2.58 then we reject the null hypothesis.

In [22]:
# define a function to calculate the Z-test statistic
# pass the population mean, population standard deviation, sample size and sample mean as the function input
def z_test(pop_mean, pop_std, n, samp_mean):

    # calculate the test statistic
    z_score = (samp_mean - pop_mean) / (pop_std / np.sqrt(n))

    # return the z-test value
    return z_score

# given data
n = 45
pop_mean = 25
pop_std = 2.5
samp_mean = 24

# calculate the test statistic using the function 'z_test'
z_score = z_test(pop_mean, pop_std, n, samp_mean)
print("Z-score:", z_score)

Z-score: -2.6832815729997477


In [24]:
# calculate the corresponding p-value for the test statistic
# use 'cdf()' to calculate P(Z <= z_score)
p_value = stats.norm.cdf(z_score)

# for a two-tailed test multiply the p-value by 2
req_p = p_value*2
print('p-value:', req_p)

p-value: 0.007290358091535638


In [ ]:
# calculate the 99% confidence interval for the population mean
# pass the sample mean to the parameter, 'loc'
# pass the scaling factor (pop_std / n^(1/2)) to the parameter, 'scale'
print('Confidence interval:', stats.norm.interval(0.99, loc = samp_mean, scale = pop_std / np.sqrt(n)))

Confidence interval: (23.040045096471452, 24.959954903528548)


Here the z score is less than -2.58, p-value is less than 0.01, also the confidence interval do not contain the value in the null hypothesis (i.e. 25), thus we reject the null hypothesis and conclude that there is enough evidence to claim that the mean mileage is different from 25kmph.

#### 2. The average calories in a slice bread of the brand 'Alphas' are 82 with a standard deviation of 15. An experiment is conducted to test the claim of the dietitians that the calories in a slice of bread are not as per the manufacturer's specification. A sample of 40 slices of bread is taken and the mean calories recorded are 95. Test the claim of dietitians with ⍺ value (significance level) as 0.05. (assume the normality of the data).

The null and alternative hypothesis is:

H<sub>0</sub>: $\mu = 82 $<br>
H<sub>1</sub>: $\mu ≠ 82 $

Here ⍺ = 0.05, for a two-tailed test calculate the critical z-value.

In [27]:
# calculate the z-value for 95% of confidence level
# use 'stats.norm.isf()' to find the z-value corresponding to the upper tail probability 'q'
# pass the value of 'alpha/2' to the parameter 'q', here alpha = 0.05
# use 'round()' to round-off the value to 2 digits
z_val = np.abs(round(stats.norm.isf(q = 0.05/2), 2))

print('Critical value for two-tailed Z-test:', z_val)

Critical value for two-tailed Z-test: 1.96


i.e. if z is less than -1.96 or greater than 1.96 then we reject the null hypothesis.

In [29]:
# define a function to calculate the Z-test statistic
# pass the population mean, population standard deviation, sample size and sample mean as the function input
def z_test(pop_mean, pop_std, n, samp_mean):

    # calculate the test statistic
    z_score = (samp_mean - pop_mean) / (pop_std / np.sqrt(n))

    # return the z-test value
    return z_score

# given data
n = 40
pop_mean = 82
pop_std = 15
samp_mean = 95

# calculate the test statistic using the function 'z_test'
z_score = z_test(pop_mean, pop_std, n, samp_mean)
print("Z-score:", z_score)

Z-score: 5.4812812776251905


In [31]:
# calculate the corresponding p-value for the test statistic
# use 'cdf()' to calculate P(Z <= -z_score)
p_value = stats.norm.cdf(-z_score)

# for a two-tailed test multiply the p-value by 2
req_p = p_value*2
print('p-value:', req_p)

p-value: 4.222565249683579e-08


In [33]:
# calculate the 95% confidence interval for the population mean
# pass the sample mean to the parameter, 'loc'
# pass the scaling factor (pop_std / n^(1/2)) to the parameter, 'scale'
print('Confidence interval:', stats.norm.interval(0.95, loc = samp_mean, scale = pop_std / np.sqrt(n)))

Confidence interval: (90.35153725771579, 99.64846274228421)


Here the z score is greater than 1.96, the p-value is less than 0.05, also the confidence interval does not contain the null value (i.e. 82), thus we reject the null hypothesis and our conclusion is, there is enough evidence to conclude that the average calories in a slice bread are not 82.

#### 3. A typhoid vaccine in the market inscribes 3 mg of ascorbic acid in the vaccine. A research team claims that the vaccines contain less than 3 mg of acid. We collected the data of 40 vaccines by using random sampling from a population and recorded the amount of ascorbic acid. Test the claim of the research team using the sample data ⍺ value (significance level) to 0.05.

    acid_amt = [2.57, 3.06, 3.28 , 3.24, 2.79, 3.40, 3.36, 3.07, 2.46, 3.03, 3.05, 2.94, 3.46, 3.19, 3.09, 2.81, 3.13, 2.88,
                2.76, 2.75, 3.17, 2.89, 2.54, 3.18, 3.08, 2.60, 3.06, 3.13, 3.11, 3.08, 2.93, 2.90, 3.06, 2.97, 3.24, 2.86,
                2.87, 3.18, 3, 2.95]

Let us check the normality of the data

In [ ]:
# given data
acid_amt = [2.57, 3.06, 3.28 , 3.24, 2.79, 3.40, 3.36, 3.07, 2.46, 3.03, 3.05, 2.94, 3.46 , 3.19, 3.09,2.81, 3.13, 2.88, 2.76,
            2.75, 3.17, 2.89, 2.54, 3.18, 3.08, 2.60, 3.06, 3.13, 3.11, 3.08, 2.93, 2.90, 3.06, 2.97, 3.24, 2.86, 2.87, 3.18,
            3, 2.95]

# perform Shapiro-Wilk test to test the normality
# shapiro() returns a tuple having the values of test statistics and the corresponding p-value
# pass the sample data to perform the test
stat, p_value = shapiro(acid_amt)

# print the test statistic and corresponding p-value
print('Test statistic:', stat)
print('P-Value:', p_value)

From the above result, we can see that the p-value is greater than 0.05, thus we can say that the data is normally distributed.

**Please note that here since n>30 we need not check for normality of the data**

The null and alternative hypothesis is:

H<sub>0</sub>: $\mu \geq 3 $<br>
H<sub>1</sub>: $\mu < 3 $

Here ⍺ = 0.05, for a one-tailed test calculate the critical z-value.

In [ ]:
# calculate the z-value for 95% of confidence level
# use 'stats.norm.isf()' to find the z-value corresponding to the upper tail probability 'q'
# pass the value of 'alpha' to the parameter 'q', here alpha = 0.05
# use 'round()' to round-off the value to 2 digits
z_val = np.abs(round(stats.norm.isf(q = 0.05), 2))

print('Critical value for one-tailed Z-test:', z_val)

i.e. if z is less than -1.64 then we reject the null hypothesis.

In [ ]:
# given amount of ascorbic acid in the vaccines
acid_amt = [2.57, 3.06, 3.28 , 3.24, 2.79, 3.40, 3.36, 3.07, 2.46, 3.03, 3.05, 2.94, 3.46 , 3.19, 3.09, 2.81, 3.13, 2.88, 2.76,
            2.75, 3.17, 2.89, 2.54, 3.18, 3.08, 2.60, 3.06, 3.13, 3.11, 3.08, 2.93, 2.90, 3.06, 2.97, 3.24, 2.86, 2.87, 3.18,
            3, 2.95]

# 'ztest()' returns the test statistic and corresponding p-value
# pass the sample data to the parameter, 'x1'
# pass the value in null hypothesis to the parameter, 'value'
# pass the one-tailed condition to the parameter, 'alternative'
z_score, pval = stests.ztest(x1 = acid_amt, value = 3, alternative = 'smaller')

# print the test statistic and corresponding p-value
print("Z-score: ", z_score)
print("p-value: ", pval)

In [ ]:
# calculate the 95% confidence interval for the population mean
# pass the sample mean to the parameter, 'loc'
# pass the scaling factor (standard deviation of sample / n^(1/2)) to the parameter, 'scale'
# use 'stdev()' to calculate sample standard deviation
print('Confidence interval:', stats.norm.interval(0.95, loc = np.mean(acid_amt),
                                                  scale = statistics.stdev(acid_amt) / np.sqrt(len(acid_amt))))

Here the z-score is greater than -1.64, the p-value is greater than 0.05, thus we fail to reject the null hypothesis. Thus, there is not enough evidence to conclude that on average a typhoid vaccine contains less than 3 mg of ascorbic acid.

**Please note that when null hypothesis is not rejected it cannot be accepted as true.There is merely not enough evidence to say that it is false. This may seem confusing but the situation is analogous to the court of law, where the verdict is either guilty or not guilty based on the evidence. If the person is judged not guilty then it does not mean the person is proved innocent but merely that there is not enough evidence to reach the guilty verdict.**

#### 4. A sample of 900 PVC pipes is found to have an average thickness of 12.5 mm. Can we assume that the sample is coming from a normal population with mean 13 mm against that it is less than 13 mm. The population standard deviation is 1 mm. Test the hypothesis at 5% level of significance.

The null and alternative hypothesis is:

H<sub>0</sub>: $\mu \geq 13 $<br>
H<sub>1</sub>: $\mu < 13 $

Here ⍺ = 0.05, for a one-tailed test calculate the critical z-value.

In [ ]:
# calculate the z-value for 95% of confidence level
# use 'stats.norm.isf()' to find the z-value corresponding to the upper tail probability 'q'
# pass the value of 'alpha' to the parameter 'q', here alpha = 0.05
# use 'round()' to round-off the value to 2 digits
z_val = np.abs(round(stats.norm.isf(q = 0.05), 2))

print('Critical value for one-tailed Z-test:', z_val)

i.e. if z is less than -1.64 then we reject the null hypothesis.

In [ ]:
# define a function to calculate the Z-test statistic
# pass the population mean, population standard deviation, sample size and sample mean as the function input
def z_test(pop_mean, pop_std, n, samp_mean):

    # calculate the test statistic
    z_score = (samp_mean - pop_mean) / (pop_std / np.sqrt(n))

    # return the z-test value
    return z_score

# given data
n = 900
pop_mean = 13
pop_std = 1
samp_mean = 12.5

# calculate the test statistic using the function 'z_test'
z_score = z_test(pop_mean, pop_std, n, samp_mean)
print("Z-score:", z_score)

In [ ]:
# calculate the corresponding p-value for the test statistic
# use 'cdf()' to calculate P(Z <= z_score)
p_value = stats.norm.cdf(z_score)

print('p-value:', p_value)

Here the z score is less than -1.64, p-value is less than 0.05, thus we reject the null hypothesis and conclude that the sample is not coming from the population with a mean less than 13.

#### 5. An e-commerce company claims that the mean delivery time of food items on its website in NYC is 60 minutes with a standard deviation of 30 minutes. A random sample of 45 customers ordered from the website, and the average time for delivery was found to be 75 minutes. Is this enough evidence to claim that the average time to get items delivered is more than 60 minutes. (assume the normality of the data). Test the with α = 0.05.

The null and alternative hypothesis is:

H<sub>0</sub>: $\mu \leq  60 $<br>
H<sub>1</sub>: $\mu > 60 $

Here ⍺ = 0.05, for a one-tailed test calculate the critical z-value.

In [ ]:
# calculate the z-value for 95% of confidence level
# use 'stats.norm.isf()' to find the z-value corresponding to the upper tail probability 'q'
# pass the value of 'alpha' to the parameter 'q', here alpha = 0.05
# use 'round()' to round-off the value to 2 digits
z_val = np.abs(round(stats.norm.isf(q = 0.05), 2))

print('Critical value for one-tailed Z-test:', z_val)

i.e. if z is greater than 1.64 then we reject the null hypothesis.

In [ ]:
# define a function to calculate the Z-test statistic
# pass the population mean, population standard deviation, sample size and sample mean as the function input
def z_test(pop_mean, pop_std, n, samp_mean):

    # calculate the test statistic
    z_score = (samp_mean - pop_mean) / (pop_std / np.sqrt(n))

    # return the z-test value
    return z_score

# given data
n = 45
pop_mean = 60
pop_std = 30
samp_mean = 75

# calculate the test statistic using the function 'z_test'
z_score = z_test(pop_mean, pop_std, n, samp_mean)
print("Z-score:", z_score)

In [ ]:
# calculate the corresponding p-value for the test statistic
# use 'sf()' to calculate P(Z > z_score)
p_value = stats.norm.sf(z_score)

print('p-value:', p_value)

Here the z score is greater than 1.64, the p-value is less than 0.05 thus we reject the null hypothesis and conclude that the average time to get items delivered is more than 60 minutes.

<a id="t"></a>
# 6. Small Sample Test

If the sample size is small (usually, n < 30) then we use the `t-test`. These tests are also known as `exact tests`.

<a id="1t"></a>
## 6.1 One Sample t Test

Let us perform a one sample t-test for the population mean. We compare the population mean with a specific value.

The null and alternative hypothesis is given as:

<p style='text-indent:25em'> <strong> $H_{0}: \mu = \mu_{0}$ or $\mu \geq \mu_{0}$ or $\mu \leq \mu_{0}$</strong></p>
<p style='text-indent:25em'> <strong> $H_{1}: \mu \neq \mu_{0}$ or $\mu < \mu_{0}$ or $\mu > \mu_{0}$</strong></p>

The test statistic is given as:
<p style='text-indent:25em'> <strong> $t = \frac{\overline{X} -  \mu_{0}}{\frac{s}{\sqrt(n)}}$</strong></p>

Where, <br>
$\overline{X}$: Sample mean<br>
$s$: Sample standard deviation<br>
$n$: Sample size

Under $H_{0}$ the test statistic follows a t-distribution with n-1 degrees of freedom.

### Example:


#### 1. A survey claims that in a math test female students tend to score fewer marks than the average marks of 75 out of 100. Consider a sample of 24 female students and perform a hypothesis test to check the claim with 90% confidence.

Use the dataset available in the CSV file `mathscore_1ttest.csv`.

In [ ]:
h0:u>75
h1:u<75

In [123]:
# read the students performance data
import pandas as pd
df_female_scores = pd.read_csv('mathscore_1ttest.csv')


# display the first two observations
len(df_female_scores)

24

In [125]:
df_female_scores=df_female_scores[df_female_scores["gender"]=="female"]["math score"]

In [43]:
female_score=df_female_scores[df_female_scores["gender"]=="female"]["math score"]

In [25]:
import numpy as np
n=24
math_score=(df_female_scores["math score"])
len(math_score)
np.mean(math_score)
np.std(math_score)
dof=n-1
dof=23
pop_mean=75

In [37]:
t_val = np.abs(round(stats.t.isf(q = 0.1,df=23), 2))

print('Critical value for one-tailed t-test:', t_val)

Critical value for one-tailed t-test: 1.32


In [41]:
# consider a list of math scores of female students from the data
math_marks = list(df_female_scores['math score'])

# calculate sample mean
sample_avg = np.mean(math_marks)

# calculate sample standard deviation
sample_std = statistics.stdev(math_marks)

# sample size
n = len(math_marks)

# degrees of freedom for 1 sample t-test
print('Degrees of freedom:', n - 1)

Degrees of freedom: 23


Note: There is no point in checking the normality of the sample. So we to rely on an underlying assumption that the variable is approximately normally distributed.

The null and alternative hypothesis is:

H<sub>0</sub>: $\mu \geq 75$<br>
H<sub>1</sub>: $\mu < 75$

Here ⍺ = 0.1 and degrees of freedom = 23, for a one-tailed test let us calculate the critical t-value.

In [43]:
# calculate the t-value for 90% of confidence level
# use 'stats.t.isf()' to find the t-value corresponding to the upper tail probability 'q'
# pass the value of 'alpha' to the parameter 'q', here alpha = 0.1
# pass the degrees of freedom to the parameter, 'df'
# use 'round()' to round-off the value to 2 digits
t_val = round(stats.t.isf(q = 0.1, df = 23), 2)

print('Critical value for one-tailed t-test:', t_val)

Critical value for one-tailed t-test: 1.32


i.e. if t is less than -1.32 then we reject the null hypothesis.

In [47]:
# use 'ttest_1samp()' to calculate the test statistic and corresponding p-value for 2-tailed test
# pass the sample data to the parameter, 'a'
# pass the average value in the null hypothesis to the parameter, 'popmean'
t_stat, p_val = stats.ttest_1samp(a = math_marks, popmean = 75)

t_stat, p_val

(-3.6067380757023204, 0.0014853227915357337)

In our example, the hypothesis test is one-tailed. Thus, we divide the two-tailed probability by 2 to obtain the one-tailed probability.

In [49]:
# divide the p-value by 2
req_p_val = p_val/2

# print the test statistic value and corresponding p-value
print('Test Statistic:', t_stat)
print('p-value:', req_p_val)

Test Statistic: -3.6067380757023204
p-value: 0.0007426613957678669


We can see that the test statistic value is less than -1.32, and the p-value is less than 0.1. Thus, we reject the null hypothesis and conclude that the proposed claim, i.e. in math tests female students tend to get lower than the average marks of
75 out of 100, is correct.

#### 2. A researcher is studying the growth of bacteria in waters of Lake Beach. The mean bacteria count of 100 per unit volume of water is within the safety level. The researcher collected 10 water samples of unit volume and found the mean bacteria count to be 94.8 with a sample variance of 72.66. Does the data indicate that the bacteria count is within the safety level? Test at the α = .05 level. Assume that the measurements constitute a sample from a normal population.

In [ ]:
pop_mean(100)
samp_mean=94.8
samp_variance=72.66
n=10
confidence=0.05

The null and alternative hypothesis is:

H<sub>0</sub>: $\mu \geq 100$<br>
H<sub>1</sub>: $\mu < 100$

Here ⍺ = 0.05 and degrees of freedom = 9, for a one-tailed test let us calculate the critical t-value.

In [ ]:
# calculate the t-value for 95% of confidence level
# use 'stats.t.isf()' to find the t-value corresponding to the upper tail probability 'q'
# pass the value of 'alpha' to the parameter 'q', here alpha = 0.05
# pass the degrees of freedom to the parameter, 'df'
# use 'round()' to round-off the value to 2 digits
t_val = round(stats.t.isf(q = 0.05, df = 9), 2)

print('Critical value for one-tailed t-test:', t_val)

As t-distribution is symmetric, for a left-tailed test if t is less than -1.83 then we reject the null hypothesis.

In [51]:
# define a function to calculate the t-test statistic
# pass the population mean, sample standard deviation, sample size and sample mean as the function input
def t_test(pop_mean, samp_std, n, samp_mean):

    # calculate the test statistic
    t_score = (samp_mean - pop_mean) / (samp_std / np.sqrt(n))

    # return the t-test value
    return t_score

# given data
n = 10
pop_mean = 100
samp_var = 72.66
samp_mean = 94.8

# calculate sample standard deviation
samp_std = np.sqrt(samp_var)

# calculate the test statistic using the function 't_test'
t_score = t_test(pop_mean, samp_std, n, samp_mean)
print("t-score:", t_score)

t-score: -1.9291040236750068


In [ ]:
# calculate the corresponding p-value for the test statistic
# use 'cdf()' to calculate P(t <= t_score)
# pass the degrees of freedom to the parameter, 'df'
p_value = stats.t.cdf(t_score, df = 9)

print('p-value:', p_value)

We can see that the test statistic value is less than -1.83, the p-value is less than 0.05. Thus, we reject the null hypothesis and can conclude that the average bacteria per unit volume (true mean) is within the safety levels.

<a id="error"></a>
# 6. Errors in Hypothesis Testing

Under the hytpothesis testing framework there can be two decisions; either we 'reject' the null hypotyhesis H<sub>0</sub> or 'not reject' the null hypothesis H<sub>0</sub>. Errors can still occur in a hypothesis testing scenario because the decision to reject or not reject the null hypothesis is made on the basis of data taken from a sample. Two types of errors are `Type I` and `Type II` error.

### Type I Error

This kind of error occurs when we reject the null hypothesis even if it is true. It is equivalent to a `false positive` conclusion. The maximum probability of committing a type I error is given by the value of $\alpha$, level of significance. So when $\alpha$ = 0.05 there is a 5% chance of rejecting a true null hypothesis.

### Type II Error

This kind of error occurs when we fail to reject the null hypothesis even if it is wrong. It is equivalent to a `false negative` conclusion. The probability of type II error is given by the value of $\beta$. The value of $\beta$ cannot be easily calculated. But $\alpha$ and $\beta$ are thus related that decreasing one increases the other.